data/tutorials/wf_03_metaprogramming.md

```diff
@@ -0,0 +1,444 @@
+---
+id: metaprogramming
+title: Preprocessors and PPXes
+description: >
+  An introduction to metaprogramming in OCaml, including preprocessors, ppx extensions and the ppxlib library.
+category: "guides"
+date: 2022-07-25T21:07:30-00:00
+---
+
+
+# Preprocessors and PPXes
+
+Preprocessors are programs meant to be called at compile-time, so that they
+alter the program's source code before the actual compilation. They can be very
+useful for several things, such as the inclusion of a file, conditional
+compilation, or use of macros.
+
+To start with an example, taking inspiration from the syntax from the C
+preprocessor, we could write something like that:
+
+
+```ocaml
+let compiler_system = #SYSTEM
```

Comment:
That's not the syntax of cppo and I worry that readers might try it without ready the context.
```suggestion
#ifndef SYSTEM
#define SYSTEM "unknown"
#endif
let compiler_system = SYSTEM
```

Here's snippet that work with cppo. The ifndef thing is to clarify that we're talking about `SYSTEM` being a preprocessor macro.

I'm not a fan of sections that explain using "we could have", "it would" and expect the reader to imagine a consistent world. Especially that all these efforts are wasted two paragraphs later. Also, it is long enough to obfuscate the paragraph that talk about the two kinds of preprocessors.
My second suggestion is to remove the entire intro.

Reply:
I think an intro with an example is important for people who do not know what a preprocessor is and click on the link. (OCaml is sometimes taught as a first programming language)

I changed the example to a shorter one using a PPX. I also lifted the paragraph on the two kinds of preprocessors as a separate section.

Reply:
The new example is nice!